### PR TITLE
Make matrix

### LINKED ISSUE #3: Determine optical path in each material along ray
Write a C++ function in OpenMC that fires a ray from `Position start` to `Position end` and determines the total optical thickness between those points at some single energy.

### LINKED ISSUE #9: Determine material boundaries by ray tracing in a multi-D problem with continuous geometry
In a multi-D problem with continuous geometry (aka OpenMC) and a voxelized solution:
The material boundaries need to be determined by ray tracing to determine k and {sigma_k}

### LINKED ISSUE #12: Unit tests for optical thickness calculation within OpenMC

### LINKED ISSUE #13: Apply sampling to cpp code rather than python

### LINKED ISSUE #15: Code formatting and best practices for OpenMC

### LINKED ISSUE #17: Add parallelization for the sampling method

### PR DESCRIPTION
**Pull request by damianjilk2**
# Description

Please include a summary of the change and which issue is fixed if applicable. Please also include relevant motivation and context.

Fixes # (issue)

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by damianjilk2** [comment]
Example code to access make_matrix.
```import openmc.lib
openmc.lib.init()
start = 10
end = 11
result = openmc.lib.make_matrix(start, end)
openmc.lib.finalize()
```

Before you can access this new version, you must update your build dir. Here is a sample bash script to do so where `openmc-fork` corresponds to this edited version of OpenMC.
```
cd openmc-fork/openmc/build 
cmake -DCMAKE_INSTALL_PREFIX=$HOME/openmc-fork/openmc-install ..
make
make install
```

File: openmc/lib/core.py [hunk 1]
```diff
@@ -107,6 +107,14 @@ class _SourceSite(Structure):
 _dll.openmc_sample_external_source.argtypes = [c_size_t, POINTER(c_uint64), POINTER(_SourceSite)]
 _dll.openmc_sample_external_source.restype = c_int
 _dll.openmc_sample_external_source.errcheck = _error_handler
+_dll.make_matrix.restrype = None
```
Comment (gonuke) [inline openmc/lib/core.py]:
```suggestion
_dll.make_matrix.restype = None
```

File: src/geometry_aux.cpp [hunk 2]
```diff
@@ -40,6 +42,57 @@ void update_universe_cell_count(int32_t a, int32_t b)
...
+      p.event_calculate_xs();
+      optical_thickness += distance * p.macro_xs().total;
+
+      if (boundary.surface_index == -1)
+        break;
+
+      p.move_distance(distance);
+    }
+    total_optical_thickness += optical_thickness;
+  }
```
Comment (gonuke) [inline src/geometry_aux.cpp]: Might make sense for this to be function that acts on a single start/end pair and then call it from this function

File: include/openmc/geometry_aux.h [hunk 3]
```diff
@@ -20,6 +20,10 @@ extern std::unordered_map<int32_t, std::unordered_map<int32_t, int32_t>>
 extern std::unordered_map<int32_t, int32_t> universe_level_counts;
 } // namespace model
 
+extern "C" void openmc_get_optical_thickness(Position start_voxel_min,
```
Comment (gonuke) [inline include/openmc/geometry_aux.h]: Needs some documentation in comments

File: tests/test_openmc_get_optical_thickness.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import openmc
+from openmc import RegularMesh
+import openmc.lib
+import numpy as np
+
+# Material
+material = openmc.Material(name="Hydrogen")
```
Comment (gonuke) [inline tests/test_openmc_get_optical_thickness.py]: We should implement this with fictitious cross-sections that are simple and well-known

File: openmc/lib/core.py [hunk 5]
```diff
@@ -107,6 +109,26 @@ class _SourceSite(Structure):
 _dll.openmc_sample_external_source.argtypes = [c_size_t, POINTER(c_uint64), POINTER(_SourceSite)]
 _dll.openmc_sample_external_source.restype = c_int
 _dll.openmc_sample_external_source.errcheck = _error_handler
+_dll.openmc_get_optical_thickness.restrype = None
+_dll.openmc_get_optical_thickness.argtypes = [_Position, _Position, _Position, _Position, c_int, POINTER(c_double)]
+
+def get_optical_thickness(start_voxel_min:tuple, start_voxel_max:tuple, end_voxel_min:tuple, end_voxel_max:tuple, num_rays:int):
```
Comment (gonuke) [inline openmc/lib/core.py]: Given my suggestion to have a new function that just calculates the optical thickness between one start/end pair, we should consider if it should also be exposed in the python API

File: src/geometry_aux.cpp [hunk 6]
```diff
@@ -40,6 +42,57 @@ void update_universe_cell_count(int32_t a, int32_t b)
...
+      if (!exhaustive_find_cell(p))
+        break;
+
+      BoundaryInfo boundary = distance_to_boundary(p);
+      double distance = boundary.distance;
+
+      p.event_calculate_xs();
+      optical_thickness += distance * p.macro_xs().total;
+
+      if (boundary.surface_index == -1)
```
Comment (gonuke) [inline src/geometry_aux.cpp]: We need an issue to focus on the need to improve/correct this stopping condition.

**Review by gonuke — CHANGES_REQUESTED** [review]
We should focus on getting this merged in the next week or 2 - and then propose a PR to openmc

**Comment by damianjilk2** [comment]
A couple of notes:

- there is a message that prints to terminal a lot (I believe whenever a new particle is made) stating "Maximum neutron transport energy: 20000000 eV for H1". This happens with strictly one nuclide per simulation and does not seem to matter which elements are used (for instance, removing the water material results in a uranium isotope in the message). I tracked the message to simulation.cpp but it seems incorrect to simply comment out the warning message to improve performance.
Perhaps this will be fixed with fictitious cross section data. 

- still looking for a way to add weird cells and verify symmetric matrix. `distance_to_boundary(particle)` is doing a lot of the logical heavy lifting but a more thorough verification should be done to ensure gaps in materials are handled as expected.

- previously had a statement disallowing `surface index = -1`. After debugging, I found out that this limitation prevented tau from being symmetric. Still uncertain what this means however. I predict it could be related to incorrectly setting up materials within the .py test but I have not looked into it yet.

File: tests/test_openmc_get_optical_thickness.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+geometry.export_to_xml()
+
+# Settings
+settings = openmc.Settings()
+settings.particles = 1
+settings.batches = 1
+settings.export_to_xml()
+
+def calculate_optical_thickness_for_voxels(mesh: RegularMesh, num_rays: int):
+    openmc.lib.init()
```
Comment (pshriwise) [inline tests/test_openmc_get_optical_thickness.py]:
```suggestion
    openmc.lib.init()
    openmc.lib.simiulation_init()
```

File: src/geometry_aux.cpp [hunk 8]
```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
...
+  int num_rays, double* output)
+{
+  double total_optical_thickness = 0.0;
+  SpatialBox start_box(start_voxel_min, start_voxel_max);
+  SpatialBox end_box(end_voxel_min, end_voxel_max);
+
+  int64_t id = 1;
+  uint64_t seed = init_seed(id, STREAM_SOURCE);
+
+  initialize_data();
```
Comment (pshriwise) [inline src/geometry_aux.cpp]:
```suggestion
```

**Review by pshriwise — COMMENTED** [review]
As discussed in-person, calling `openmc.lib.simulation_init` should take care of the warning message you were seeing @damianjilk2.

File: openmc/lib/core.py (hunk 5 above)
Comment (gonuke) [reply openmc/lib/core.py]: I don't think I was clear about this.  You now have `caluclate_optical_thickness` and `get_optical_thickness`. I was suggesting that make both available via the `_dll` - not necessarily that you write a python method (although that may be good, too)

File: src/geometry_aux.cpp [hunk 9]
```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
   }
 }
 
+double calculate_optical_thickness(Position start_sampled_position,
+  Position end_sampled_position, uint64_t* seed)
```
Comment (gonuke) [inline src/geometry_aux.cpp]: you don't need a `seed` here
```suggestion
  Position end_sampled_position)
```

File: src/geometry_aux.cpp [hunk 10]
```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
...
+  p.from_source(&site);
+
+  double optical_thickness = 0.0;
+
+  while (true) {
+    if (!exhaustive_find_cell(p))
+      break;
+
+    BoundaryInfo boundary = distance_to_boundary(p);
+    double distance_to_boundary = boundary.distance;
```
Comment (gonuke) [inline src/geometry_aux.cpp]: Don't need to store this `Boundary` object
```suggestion
    double distance_to_boundary = distance_to_boundary(p).distance;
```

File: src/geometry_aux.cpp [hunk 11]
```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
...
+    double distance_to_endpoint = (end_sampled_position - p.r()).norm();
+
+    // Check if we are in the cell containing the endpoint
+    if (distance_to_endpoint < distance_to_boundary) {
+      // Use the remaining distance to the endpoint
+      p.event_calculate_xs();
+      optical_thickness += distance_to_endpoint * p.macro_xs().total;
+      break; // Stop the loop
+    }
```
Comment (gonuke) [inline src/geometry_aux.cpp]: You are doing the `event_calcualte_xs` as well as the incrementing of `optical_thickness` both in this if clause and if you don't follow it.  The logic will be simpler if you move this check to after the ones below and only break if necessary.

Really, this should be an if/else for break vs `move_distance()`

File: src/geometry_aux.cpp [hunk 12]
```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
...
+  while (true) {
+    if (!exhaustive_find_cell(p))
+      break;
+
+    BoundaryInfo boundary = distance_to_boundary(p);
+    double distance_to_boundary = boundary.distance;
+    double distance_to_endpoint = (end_sampled_position - p.r()).norm();
+
+    // Check if we are in the cell containing the endpoint
+    if (distance_to_endpoint < distance_to_boundary) {
```
Comment (gonuke) [inline src/geometry_aux.cpp]: This stopping condition is a good start.  We'll have to think through some of the pathologies that may occur and write tests to confirm that it works over all those edge cases.

File: src/geometry_aux.cpp [hunk 13]
```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
...
+    p.event_calculate_xs();
+    optical_thickness += distance_to_boundary * p.macro_xs().total;
+
+    p.move_distance(distance_to_boundary);
+  }
+
+  return optical_thickness;
+}
+
+extern "C" void openmc_get_optical_thickness(Position start_voxel_min,
```
Comment (gonuke) [inline src/geometry_aux.cpp]: This function and the previous one have similar names now.  Maybe this one can be renamed to better convey what it's doing.  `get_mean_optical_thickness_between_voxels` or something like that?

**Review by gonuke — CHANGES_REQUESTED** [review]
The refactor here is good, some follow-on comments to bring this home.

File: openmc/lib/core.py (hunk 5 above)
Comment (damianjilk2) [reply openmc/lib/core.py]: I will make it accessible to python api as well. We can change after discussing use cases further to determine what should be available

File: src/geometry_aux.cpp (hunk 11 above)
Comment (damianjilk2) [reply src/geometry_aux.cpp]: Good point. However, I don't think it will clean up quite as nice as hoped for since adding to the optical thickness will vary depending on using `distance_to_endpoint` or `distance_to_boundary`

**Comment by damianjilk2** [comment]
Optical thickness calculation generated from test file (10,000 rays)

  | Voxel 0 | Voxel 1 | Voxel 2 | Voxel 3 | Voxel 4 | Voxel 5 | Voxel 6 | Voxel 7
-- | -- | -- | -- | -- | -- | -- | -- | --
Voxel 0 | 1.284555 | 1.643602 | 1.64283 | 2.055349 | 1.642064 | 2.054638 | 2.053499 | 2.397973
Voxel 1 | 1.649533 | 0.814222 | 1.882247 | 1.432811 | 1.880309 | 1.432797 | 2.244048 | 1.878883
Voxel 2 | 1.651829 | 1.884811 | 0.814222 | 1.434143 | 1.881751 | 2.246266 | 1.432797 | 1.880147
Voxel 3 | 2.062959 | 1.436233 | 1.435014 | 0.814222 | 2.244596 | 1.881751 | 1.880309 | 1.432797
Voxel 4 | 1.649182 | 1.884225 | 1.883178 | 2.24704 | 0.814222 | 1.434143 | 1.432811 | 1.880851
Voxel 5 | 2.0646 | 1.436183 | 2.247067 | 1.883178 | 1.435014 | 0.814222 | 1.882247 | 1.432811
Voxel 6 | 2.065278 | 2.248764 | 1.436183 | 1.884225 | 1.436233 | 1.884811 | 0.814222 | 1.434143
Voxel 7 | 2.409221 | 1.884315 | 1.884408 | 1.436183 | 1.882323 | 1.436233 | 1.435014 | 0.814222

**Review by lewisgross1296 — COMMENTED** [review]
`test_openmc_get_optical_thickness.py` should probably live in `tests/unit_tests`, or `tests/regression_tests` if not unit tests. Consider looking at [this test](https://github.com/openmc-dev/openmc/blob/develop/tests/unit_tests/test_geometry.py) for some guidance

File: src/geometry.cpp [hunk 14]
```diff
@@ -490,4 +491,62 @@ extern "C" int openmc_global_bounding_box(double* llc, double* urc)
   return 0;
 }
 
+extern "C" double openmc_calculate_optical_thickness(
+  Position start_sampled_position, Position end_sampled_position)
```
Comment (gonuke) [inline src/geometry.cpp]: Not only does this match your declaration in the header, there may be use cases for this where the positions aren't "sampled"
```suggestion
  Position start_pos, Position end_pos)
```
(and rename all subsequent uses)

File: src/geometry.cpp [hunk 15]
```diff
@@ -490,4 +491,62 @@ extern "C" int openmc_global_bounding_box(double* llc, double* urc)
...
+  site.E = 1.0;
+  site.particle = ParticleType::neutron;
+  site.r = start_sampled_position;
+  site.u = direction;
+  p.from_source(&site);
+
+  double optical_thickness = 0.0;
+
+  while (true) {
+    if (!exhaustive_find_cell(p))
```
Comment (gonuke) [inline src/geometry.cpp]: It's inefficient to find the cell each time you cross a surface.  I think you will want to call `Particle::cross_surface()` at the appropriate spot

File: src/geometry.cpp [hunk 16]
```diff
@@ -490,4 +491,62 @@ extern "C" int openmc_global_bounding_box(double* llc, double* urc)
...
+    double endpoint_distance = (end_sampled_position - p.r()).norm();
+
+    p.event_calculate_xs();
+    if (endpoint_distance < cell_boundary_distance) {
+      optical_thickness += endpoint_distance * p.macro_xs().total;
+      break; // reached endpoint
+    } else {
+      optical_thickness += cell_boundary_distance * p.macro_xs().total;
+      p.move_distance(cell_boundary_distance);
+    }
```
Comment (gonuke) [inline src/geometry.cpp:534]: This may be where you call `p.cross_surface()`

File: src/geometry.cpp [hunk 17]
```diff
@@ -490,4 +491,62 @@ extern "C" int openmc_global_bounding_box(double* llc, double* urc)
...
+  SourceSite site;
+  site.E = 1.0;
+  site.particle = ParticleType::neutron;
+  site.r = start_sampled_position;
+  site.u = direction;
+  p.from_source(&site);
+
+  double optical_thickness = 0.0;
+
+  while (true) {
```
Comment (gonuke) [inline src/geometry.cpp]: It may be possible to set this up as a while loop for `end_point_distance > cell_boundary_distance`

File: tests/unit_tests/test_optical_thickness.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    mat1_xs = 0.01
+    mat2_xs = 0.02
+    mat1_r = 2
+    mat2_r = 10
+
+    # Origin to outer boundary in x-direction
+    expected_tau_x = mat1_xs * mat1_r + mat2_xs * (mat2_r - mat1_r)
+    calculated_tau_x = openmc.lib.calculate_optical_thickness((0, 0, 0), (10, 0, 0))
+    assert np.isclose(calculated_tau_x, expected_tau_x), \
+        f"Expected {expected_tau_x}, but got {calculated_tau_x}"
```
Comment (gonuke) [inline tests/unit_tests/test_optical_thickness.py]: Don't need these strings - pytest should report discrepancies verbosely

File: tests/unit_tests/test_optical_thickness.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+def lib_simulation_init(lib_init):
+    openmc.lib.simulation_init()
+    yield
+
+def test_optical_thickness_variations(lib_init):
+    """Test optical thickness calculation for various scenarios."""
+    mat1_xs = 0.01
+    mat2_xs = 0.02
+    mat1_r = 2
+    mat2_r = 10
```
Comment (gonuke) [inline tests/unit_tests/test_optical_thickness.py]: Can these be defined in a test fixture and used where necessary throughout? That will allow you to try different values consistently

**Review by gonuke — CHANGES_REQUESTED** [review]
This is coming together nicely.

File: src/geometry.cpp (hunk 14 above)
Comment (damianjilk2) [reply src/geometry.cpp]: hmm. I wonder why it still runs properly with the header discrepancy. I wonder if it only considers the types and method name rather than the parameter names

File: src/geometry.cpp (hunk 17 above)
Comment (damianjilk2) [reply src/geometry.cpp]: initializing variables beforehand and setting logic `>=` should successfully implement this

**Review by gonuke — APPROVED** [review]
I haven't run the tests, but I think this looks good.
